include/aspect/material_model/averaging.h

```diff
@@ -0,0 +1,200 @@
+         */
+        virtual void
+        parse_parameters (ParameterHandler &prm);
+
+        /**
+         * Method to determine viscosity dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        viscosity_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine density dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        density_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine compressibility dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        compressibility_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine specific heat capacity dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        specific_heat_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine thermal conductivity dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        thermal_conductivity_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method that indicates whether material is compressible. Depth dependent model is compressible
+         * if and only if base model is compressible.
+         */
+        virtual bool is_compressible () const;
+
+        /**
+         * Method to calculate reference viscosity for the depth-dependent model. The reference
+         * viscosity is determined by evaluating the depth-dependent part of the viscosity at
+         * the mean depth of the model.
+         */
+        virtual double reference_viscosity () const;
+
+        /**
+         * Method to calculate reference density for the depth-dependent model. Because the depth-
+         * dependent model deos not modify density, the reference density is equivalent to the
+         * base model's reference density.
+         */
+        virtual double reference_density () const;
+
+
+
+      private:
+
+        /**
+               * Parse a string representing one of the options returned by
+               * get_averaging_operation_names(), and return the corresponding
+               * AveragingOperation value.
+               */
+              AveragingOperation
+			  parse_averaging_operation_name (const std::string &s);
+
+        /**
+        * Given the averaging @p operation, a description of where the
+        * quadrature points are located on the given cell, and a mapping,
+        * perform this operation on all elements of the @p values structure.
+        */
+        void
+        average (const AveragingOperation &averaging_operation,
```

Comment:
Passing the operation by reference does not harm, but also doesn't help very much for integer types. Leave it const but remove the reference.